pkg/dependencies/factory.go

```diff
@@ -253,19 +255,21 @@ func (f *Factory) WithGovc() *Factory {
 	return f
 }
 
-func (f *Factory) WithTink(clusterConfigFile string) *Factory {
+func (f *Factory) WithTinkerbellClients(clusterConfigFile string) *Factory {
 	f.WithExecutableBuilder()
 
 	f.buildSteps = append(f.buildSteps, func(ctx context.Context) error {
-		if f.dependencies.Tink != nil {
-			return nil
-		}
 		tinkerbellDatacenterConfig, err := v1alpha1.GetTinkerbellDatacenterConfig(clusterConfigFile)
 		if err != nil {
 			return err
 		}
-		f.dependencies.Tink = f.executableBuilder.BuildTinkExecutable(tinkerbellDatacenterConfig.Spec.TinkerbellCertURL, tinkerbellDatacenterConfig.Spec.TinkerbellGRPCAuth)
+		tinkClient := f.executableBuilder.BuildTinkExecutable(tinkerbellDatacenterConfig.Spec.TinkerbellCertURL, tinkerbellDatacenterConfig.Spec.TinkerbellGRPCAuth)
 
+		pbnjClient, err := pbnj.NewPBNJClient(tinkerbellDatacenterConfig.Spec.TinkerbellPBnJGRPCAuth)
```

Comment:
I think this should be broken down into atomic constructs, i.e. 1 for the tink executable and 1 for pbnj. At present it forces consumers to create both even if they don't want them. 

I'd even like to propose a way to get away from this construct all together as its basically got knowledge of how to construct everything in the program which is a lot of responsibility.

Reply:
Actually it was originally atomic, but we wanted to make it combined as they are tinkerbell specific dependencies and would be needed together.
Also, hopefully in the near future we can eliminate the need for these 2 clients as everything would become `kubectl` driven.

And I think there have been discussions on maybe refactoring this approach, like reducing responsibility of the dependencies factory as its becoming hard to maintain. We could discuss on that more :)

Reply:
Glad you chimed in, I saw your other comment. Whether or not the provider accepts individual parameters or a single struct for clients is orthogonal to how they're constructed. I.e. you can construct the individual clients as separate entities, then combine them when the provider is created. 

If I paint a different picture for you, there could be consuming constructs that want _only_ a Tink client or _only_ a PBnJ client. Having individual constructor functions with a _single responsibility_ enables that world. The current approach here _couples_ creating a Tink client with a PBnJ client when they are _logically distinct_ and 1 doesn't strictly need the other.

Reply:
Re refactoring of `dependencies`, yes that sounds good - lets chat at some point.

Reply:
Ah I see what you meant. I’m fine with breaking down the factory methods and then injecting them with a parent struct.

My original intention was to have a wrapper struct that represents both the interfaces that can be injected into the provider. Just to simplify the number of individual objects being passed in.

Reply:
There would still be a struct for the provider representing clients. But, as I said, that's orthogonal to how those clients are constructed. I spoke with Pankti and pointed out the `WithProviderFactory()` function l. 185 can create the client struct inline. The point there is that the struct is a property of the provider factory and it doesn't need to impact the construction of the individual clients. 

Happy to chat to tomorrow to clarify further.